### PR TITLE
Use only one shard with a single thread

### PR DESCRIPTION
**Pull request by Zoxc**
This changes `Sharded` to only access a single shard using a mask set to `0` when a single thread is used, which leads to cache utilization improvements. 

Performance improvement with 1 thread and `cfg(parallel_compiler)`:
<table><tr><td rowspan="2">Benchmark</td><td colspan="1"><b>Before</b></th><td colspan="2"><b>After</b></th></tr><tr><td align="right">Time</td><td align="right">Time</td><td align="right">%</th></tr><tr><td>🟣 <b>clap</b>:check</td><td align="right">1.7402s</td><td align="right">1.7004s</td><td align="right">💚  -2.29%</td></tr><tr><td>🟣 <b>hyper</b>:check</td><td align="right">0.2633s</td><td align="right">0.2550s</td><td align="right">💚  -3.12%</td></tr><tr><td>🟣 <b>regex</b>:check</td><td align="right">0.9716s</td><td align="right">0.9482s</td><td align="right">💚  -2.41%</td></tr><tr><td>🟣 <b>syn</b>:check</td><td align="right">1.5679s</td><td align="right">1.5358s</td><td align="right">💚  -2.05%</td></tr><tr><td>🟣 <b>syntex_syntax</b>:check</td><td align="right">6.0569s</td><td align="right">5.9272s</td><td align="right">💚  -2.14%</td></tr><tr><td>Total</td><td align="right">10.5999s</td><td align="right">10.3666s</td><td align="right">💚  -2.20%</td></tr><tr><td>Summary</td><td align="right">1.0000s</td><td align="right">0.9760s</td><td align="right">💚  -2.40%</td></tr></table>

cc @SparrowLii

### CODE REVIEW

**Comment by SparrowLii** [comment]
`Sharded` uses `Lock` internal. I'm thinking if it uses `Mutex` when `is_dyn_thread_safe() == true` and `RefCell` otherwise directly, will it be faster?

File: compiler/rustc_data_structures/src/sharded.rs [hunk 1]
```diff
@@ -31,31 +35,53 @@ impl<T: Default> Default for Sharded<T> {
...
+    #[inline(always)]
+    fn count(&self) -> usize {
+        self.mask() + 1
     }
 
     /// The shard is selected by hashing `val` with `FxHasher`.
     #[inline]
     pub fn get_shard_by_value<K: Hash + ?Sized>(&self, val: &K) -> &Lock<T> {
-        if SHARDS == 1 { &self.shards[0].0 } else { self.get_shard_by_hash(make_hash(val)) }
+        self.get_shard_by_hash(if SHARDS == 1 { 0 } else { make_hash(val) })
```
Comment (SparrowLii) [inline compiler/rustc_data_structures/src/sharded.rs:65]: Should we return `&self.shards.get_unchecked(0).0` directly if `SHARDS == 1`?

Comment (Zoxc) [reply compiler/rustc_data_structures/src/sharded.rs:65]: It gets masked by 0 later, so it doesn't matter.

File: compiler/rustc_data_structures/src/sharded.rs [hunk 2]
```diff
@@ -31,31 +35,53 @@ impl<T: Default> Default for Sharded<T> {
 impl<T> Sharded<T> {
     #[inline]
     pub fn new(mut value: impl FnMut() -> T) -> Self {
-        Sharded { shards: [(); SHARDS].map(|()| CacheAligned(Lock::new(value()))) }
+        Sharded {
+            #[cfg(parallel_compiler)]
+            mask: if is_dyn_thread_safe() { SHARDS - 1 } else { 0 },
```
Comment (nnethercote) [inline compiler/rustc_data_structures/src/sharded.rs:40]: `mask` needs a comment explaining what it's for.

File: compiler/rustc_data_structures/src/sharded.rs [hunk 3]
```diff
@@ -31,31 +35,53 @@ impl<T: Default> Default for Sharded<T> {
...
+        }
+        #[cfg(not(parallel_compiler))]
+        {
+            0
+        }
+    }
+
+    #[inline(always)]
+    fn count(&self) -> usize {
+        self.mask() + 1
```
Comment (cjgillot) [inline compiler/rustc_data_structures/src/sharded.rs:59]: This needs a reminder that `mask` is `2^n - 1`, so adding 1 is correct.

File: compiler/rustc_data_structures/src/sharded.rs (hunk 1 above)
Comment (cjgillot) [reply compiler/rustc_data_structures/src/sharded.rs:65]: Could you make this a branch over `cfg(debug_assertions)`? I'd rather avoid having some logic over the cfg and some other over the value of `SHARDS`, even if they are equivalent.

**Review by cjgillot — COMMENTED** [review]
What is the reason for the perf improvement? Is it just a cache effect, or do we gain from skipping some key hashing?

File: compiler/rustc_data_structures/src/sharded.rs (hunk 1 above)
Comment (Zoxc) [reply compiler/rustc_data_structures/src/sharded.rs:65]: Assuming you mean `cfg(parallel_compiler)` here, it's still useful to disable just sharding by setting `SHARDS` to 1 to evaluate how effective sharding is.

**Comment by Zoxc** [comment]
> What is the reason for the perf improvement? Is it just a cache effect, or do we gain from skipping some key hashing?

It probably just the cache effect and the replacement of the bounds check by a masking operation. Not sure there's any place where hashing can be skipped atm.

**Pull request merged**